src/SocketEvents.ts

```diff
@@ -0,0 +1,350 @@
+import type { ERRORS } from "./Connection";
+
+/** Defines which events are emitted by the server and can be listened on the client */
+export interface IOListenEvents {
+	objectChange: (id: string, obj: ioBroker.Object) => void;
+	stateChange: (id: string, obj: ioBroker.State) => void;
+	cmdStdout: (id: string, text: string) => void;
+	cmdStderr: (id: string, text: string) => void;
+	cmdExit: (id: string, exitCode: number) => void;
+
+	connect: (noTimeout: boolean) => void;
+	reconnect: () => void;
+	disconnect: () => void;
+	reauthenticate: () => void;
+
+	log: (message: string) => void;
+
+	error: (error: Error) => void;
+	connect_error: (error: Error) => void;
+	permissionError: (error: any) => void; // TODO: check ioBroker.admin/lib/socket.js for the shape of this
+}
+
+type GenericCallback<T> = ErrorAsString<ioBroker.GenericCallback<T>>;
```

Comment:
Please export all types (unless there is a good reason not to).
Why? In some more complex cases, one might want to write a method that then calls into this library and that method might need the type.